pilot/pkg/xds/endpoints/endpoint_builder.go

```diff
@@ -427,25 +403,123 @@ func (b *EndpointBuilder) BuildLocalityLbEndpointsFromShards(
+		// To ensure we allow traffic only to mTLS endpoints, we filter out non-mTLS endpoints for these cluster types.
+		locEps = b.EndpointsWithMTLSFilter(locEps)
+	}
+
 	return locEps
 }
 
+// addUint32AvoidOverflow returns sum of two uint32 and status. If sum overflows,
+// and returns MaxUint32 and status.
+func addUint32(left, right uint32) (uint32, bool) {
+	if math.MaxUint32-right < left {
+		return math.MaxUint32, true
+	}
+	return left + right, false
+}
+
+func (b *EndpointBuilder) filterIstioEndpoint(ep *model.IstioEndpoint, svcPort *model.Port) bool {
+	// for ServiceInternalTrafficPolicy
+	if b.service.Attributes.NodeLocal && ep.NodeName != b.proxy.GetNodeName() {
+		return false
+	}
+	// Only send endpoints from the networks in the network view requested by the proxy.
+	// The default network view assigned to the Proxy is nil, in that case match any network.
+	if !b.proxyView.IsVisible(ep) {
+		// Endpoint's network doesn't match the set of networks that the proxy wants to see.
+		return false
+	}
+	// If the downstream service is configured as cluster-local, only include endpoints that
+	// reside in the same cluster.
+	if b.clusterLocal && (b.clusterID != ep.Locality.ClusterID) {
+		return false
+	}
+	// TODO(nmittler): Consider merging discoverability policy with cluster-local
+	if !ep.IsDiscoverableFromProxy(b.proxy) {
+		return false
+	}
+	if svcPort.Name != ep.ServicePortName {
+		return false
+	}
+	// Port labels
+	if !b.subsetLabels.SubsetOf(ep.Labels) {
+		return false
+	}
+	// If we don't know the address we must eventually use a gateway address
+	if ep.Address == "" && ep.Network == b.network {
+		return false
+	}
+	// Draining endpoints are only sent to 'persistent session' clusters.
+	draining := ep.HealthStatus == model.Draining ||
+		features.DrainingLabel != "" && ep.Labels[features.DrainingLabel] != ""
+	if draining {
+		persistentSession := b.service.Attributes.Labels[features.PersistentSessionLabel] != ""
+		if !persistentSession {
+			return false
+		}
+	}
+	return true
+}
+
+// snapshotShards into a local slice to avoid lock contention
+func (b *EndpointBuilder) snapshotShards(endpointIndex *model.EndpointIndex) []*model.IstioEndpoint {
+	shards := b.findShards(endpointIndex)
+	if shards == nil {
+		return nil
+	}
+
+	// Determine whether or not the target service is considered local to the cluster
+	// and should, therefore, not be accessed from outside the cluster.
+	isClusterLocal := b.clusterLocal
+
+	var eps []*model.IstioEndpoint
+	shards.RLock()
+	// Extract shard keys so we can iterate in order. This ensures a stable EDS output.
+	keys := shards.Keys()
+	// The shards are updated independently, now need to filter and merge for this cluster
+	for _, shardKey := range keys {
+		if shardKey.Cluster != b.clusterID {
+			// If the downstream service is configured as cluster-local, only include endpoints that
+			// reside in the same cluster.
+			if isClusterLocal || b.service.Attributes.NodeLocal {
```

Comment:
duplicate with the filter later

Reply:
This is a very slight optimization allowing us to not loop over `endpoints[]` because every one of them will be a miss.